dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/video/frames_decoder_gpu.h"
+
+#include <cuda.h>
+#include <unistd.h>
+
+#include <string>
+#include <memory>
+
+#include "dali/core/error_handling.h"
+#include "dali/core/cuda_utils.h"
+#include "dali/pipeline/data/backend.h"
+#include "dali/pipeline/data/tensor.h"
+#include "dali/operators/reader/loader/video/nvdecode/color_space.h"
+
+namespace dali {
+namespace detail {
+int process_video_sequence(void *user_data, CUVIDEOFORMAT *video_format) {
+  return video_format->min_num_decode_surfaces;
+}
+
+int process_picture_decode(void *user_data, CUVIDPICPARAMS *picture_params) {
+  FramesDecoderGpu *frames_decoder = static_cast<FramesDecoderGpu*>(user_data);
+
+  // Sending empty packet will call this callback.
+  // If we want to flush the decoder, we do not need to do anything here
+  if (frames_decoder->flush_) {
+    return 0;
+  }
+
+  CUDA_CALL(cuvidDecodePicture(frames_decoder->nvdecode_state_->decoder, picture_params));
+
+  // Process decoded frame for output
+  CUVIDPROCPARAMS videoProcessingParameters = {};
+  videoProcessingParameters.progressive_frame = !picture_params->field_pic_flag;
+  videoProcessingParameters.second_field = 1;
+  videoProcessingParameters.top_field_first = picture_params->bottom_field_flag ^ 1;
+  videoProcessingParameters.unpaired_field = 0;
+  videoProcessingParameters.output_stream = frames_decoder->stream_;
+
+  uint8_t *frame_output = nullptr;
+
+  // Take pts of the currently decoded frame
+  int current_pts = frames_decoder->piped_pts_.front();
+  frames_decoder->piped_pts_.pop();
+
+  // current_pts is pts of frame that came from the decoder
+  // CurrentFramePts() is pts of the frame that we want to return
+  // in this call to ReadNextFrame
+  // If they are the same, we just return this frame
+  // If not, we store it in the buffer for later
+
+  if (current_pts == frames_decoder->NextFramePts()) {
+    // Currently decoded frame is actually the one we wanted
+    frames_decoder->frame_returned_ = true;
+    if (frames_decoder->current_copy_to_output_ == false) {
+      return 1;
+  }
```

Comment:
nitpick: formatting - should be indented?

Reply:
Done